packages/generate/src/gel-prisma.ts

```diff
@@ -0,0 +1,714 @@
+        ),
+        target: {name},
+
+        properties: {
+            name,
+            readonly,
+            required,
+            cardinality,
+            exclusive := exists (
+                select .constraints
+                filter .name = 'std::exclusive'
+            ),
+            target: {name},
+        },
+    } filter .name != '__type__',
+    properties: {
+        name,
+        readonly,
+        required,
+        cardinality,
+        exclusive := exists (
+            select .constraints
+            filter .name = 'std::exclusive'
+        ),
+        target: {name},
+    },
+    backlinks := <array<str>>[],
+}
+filter
+    not .builtin
+    and
+    not .internal
+    and
+    not re_test('^(std|cfg|sys|schema)::', .name);
+`;
+
+const MODULE_QUERY = `
+with
+    module schema,
+    m := (select \`Module\` filter not .builtin)
+select m.name;
+`;
+
+interface JSONField {
+  name: string;
+  readonly?: boolean;
+  required?: boolean;
+  exclusive?: boolean;
+  cardinality: string;
+  target: { name: string };
+  has_link_target?: boolean;
+  has_link_object?: boolean;
+}
+
+interface JSONLink extends JSONField {
+  fwname?: string;
+  properties?: JSONField[];
+}
+
+interface JSONType {
+  name: string;
+  links: JSONLink[];
+  properties: JSONField[];
+  backlinks: JSONLink[];
+  backlink_renames?: { [key: string]: string };
+}
+
+interface LinkTable {
+  module: string;
+  name: string;
+  table: string;
+  source: string;
+  target: string;
+}
+
+interface TableObj {
+  module: string;
+  name: string;
+  table: string;
+  links: JSONField[];
+  properties: JSONField[];
+}
+
+interface ProcessedSpec {
+  modules: string[];
+  object_types: JSONType[];
+  link_tables: LinkTable[];
+  link_objects: TableObj[];
+  prop_objects: TableObj[];
+}
+
+interface MappedSpec {
+  link_objects: { [key: string]: TableObj };
+  object_types: { [key: string]: JSONType };
+}
+
+const CLEAN_NAME = /^[A-Za-z_][A-Za-z0-9_]*$/;
+
+function warn(msg: string) {
+  // Keep record of warnigns during the generation process
+  WARNINGS.push(msg);
+}
+
+function getSQLName(name: string): string {
+  // Just remove the module name
+  return <string>name.split("::").pop();
+}
+
+function getMod(fullname: string): string {
+  return <string>fullname.split("::").slice(0, -1).join("::");
+}
+
+function getModAndName(fullname: string): string[] {
+  const mod = fullname.split("::");
+  const name = <string>mod.pop();
+  return [mod.join("::"), name];
+}
+
+function validName(name: string): boolean {
+  // Just remove module separators and check the rest
+  name = name.replace("::", "");
+  if (!CLEAN_NAME.test(name)) {
+    warn(`Non-alphanumeric names are not supported: ${name}`);
+    return false;
+  }
+  return true;
+}
+
+async function getSchemaJSON(client: Client) {
+  const types = <JSONType[]>JSON.parse(await client.queryJSON(INTRO_QUERY));
+  const modules = <string[]>JSON.parse(await client.queryJSON(MODULE_QUERY));
```

Comment:
Why `queryJSON` -> `JSON.parse` instead of `query`?

Reply:
This was mostly a translation from the Python version of this tool. There I was collecting things into dicts rather than native objects so that I could access things uniformly via `[ ]` and not worry about fields. I realize that in JS version this is not necessary, so a query is probably right in this spot.